helix-core/src/match_brackets.rs

```diff
@@ -1,55 +1,149 @@
-    ('(', ')'),
-    ('{', '}'),
-    ('[', ']'),
-    ('<', '>'),
-    ('\'', '\''),
-    ('"', '"'),
+    (O_PAREN, C_PAREN),
+    (O_CURLY, C_CURLY),
+    (O_BRCKT, C_BRCKT),
+    (O_ANGLE, C_ANGLE),
+    (SINGLE_QUOTE, SINGLE_QUOTE),
+    (DOUBLE_QUOTE, DOUBLE_QUOTE),
 ];
+
+const VALID_CHARS: &[char] = &[
+    O_PAREN,
+    C_PAREN,
+    O_CURLY,
+    C_CURLY,
+    O_BRCKT,
+    C_BRCKT,
+    O_ANGLE,
+    C_ANGLE,
+    SINGLE_QUOTE,
+    DOUBLE_QUOTE,
+];
+
 // limit matching pairs to only ( ) { } [ ] < >
 
+// Returns the position of the matching bracket under cursor.
+//
+// If the cursor is one the opening bracket, the position of
+// the closing bracket is returned. If the cursor in the closing
+// bracket, the position of the opening bracket is returned.
+//
+// If the cursor is not on a bracket, `None` is returned.
 #[must_use]
-pub fn find(syntax: &Syntax, doc: &Rope, pos: usize) -> Option<usize> {
+pub fn find_matching_bracket(syntax: &Syntax, doc: &Rope, pos: usize) -> Option<usize> {
     let tree = syntax.tree();
 
-    let byte_pos = doc.char_to_byte(pos);
+    if pos >= doc.len_chars() || !VALID_CHARS.contains(&doc.char(pos)) {
+        return None;
+    }
 
-    // most naive implementation: find the innermost syntax node, if we're at the edge of a node,
-    // return the other edge.
+    let byte_pos = doc.char_to_byte(pos);
 
     let node = match tree
         .root_node()
         .named_descendant_for_byte_range(byte_pos, byte_pos)
     {
-        Some(node) => node,
-        None => return None,
+        Some(node) if !node.is_error() => node,
+        _ => return None,
     };
 
-    if node.is_error() {
-        return None;
+    let (start_byte, end_byte) = surrounding_bytes(&doc, &node)?;
+    let (start_char, end_char) = (doc.byte_to_char(start_byte), doc.byte_to_char(end_byte));
+
+    if is_valid_pair(&doc, start_char, end_char) {
+        if start_byte == byte_pos {
+            return Some(end_char);
+        }
+        if end_byte == byte_pos {
+            return Some(start_char);
+        }
     }
 
-    let len = doc.len_bytes();
-    let start_byte = node.start_byte();
-    let end_byte = node.end_byte().saturating_sub(1); // it's end exclusive
-    if start_byte >= len || end_byte >= len {
-        return None;
+    None
+}
+
+// Returns the position of the bracket that is closing the current scope.
+//
+// If the cursor is on an opening or closing bracket, the function
+// behaves equivalent to [`find_matching_bracket`].
+//
+// If the cursor position is within a scope, the function searches
+// for the surrounding scope that is surrounded by brackets and
+// returns the position of the closing bracket for that scope.
+//
+// If no surrounding scope is found, the function returns `None`.
+#[must_use]
+pub fn find_matching_bracket_fuzzy(syntax: &Syntax, doc: &Rope, pos: usize) -> Option<usize> {
+    let tree = syntax.tree();
+    let byte_pos = doc.char_to_byte(pos);
+
+    let mut cursor = tree.walk();
```

Comment:
This whole block can be replaced by

```rust
    let mut node = tree
        .root_node()
        .named_descendant_for_byte_range(byte_pos, byte_pos)?;

    // Travere the tree upwards to find first node with surrounding brackets.
    loop {
        let (start_byte, end_byte) = surrounding_bytes(&doc, &node)?;
        let (start_char, end_char) = (doc.byte_to_char(start_byte), doc.byte_to_char(end_byte));

        if is_valid_pair(&doc, start_char, end_char) {
            if start_byte == byte_pos {
                return Some(end_char);
            }
            if end_byte == byte_pos {
                return Some(start_char);
            }
            // We found a surrounding node, but the cursor
            // is within the scope of that node. Hence, we
            // return the closing bracket.
            return Some(end_char);
        }

        node = node.parent()?;
    }
```

which ends up being very similar to the `find_matching_bracket` (a good refactoring opportunity :).

Reply:
I refactored it a little bit and we now use the same logic for both kinds of bracket search :)